cmd/controller-manager/main.go

```diff
@@ -54,6 +55,72 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// ensureCRDExists to ensure CRD exists before start
+func ensureCRDExists(ns string, kubeClient kubernetes.Interface, clientSet *versioned.Clientset) error {
+	condition := func() (bool, error) {
+		var err error
+		_, err = kubeClient.CoreV1().Namespaces().Get(ns, metav1.GetOptions{})
+		if err == nil {
+			// Only namespace(the controller deploy under it) exists, to list resources
+			// checking the CRD directly but that will require new rbacs, the purpose of the check is that the endpoint exists, it doesn't matter if there's nothing in the namespace
+			_, err = clientSet.PingcapV1alpha1().Backups(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list backups with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().BackupSchedules(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list backup Schedules with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().DMClusters(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list dm clusters with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().Restores(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list restores with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().TidbClusters(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list tidb clusters with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().TidbInitializers(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list tidb initializers with error=%+v", err)
+				return false, err
+			}
+
+			_, err = clientSet.PingcapV1alpha1().TidbMonitors(ns).List(metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list tidb monitors with error=%+v", err)
+				return false, err
+			}
+			return true, nil
+		}
+		klog.Errorf("Failed to get namespace %s, error=%+v", ns, err)
+		return false, err
+	}
+	// with a Factor of 1.5 we wait up to 7.5 seconds (the 10th attempt)
+	backoff := wait.Backoff{
+		Duration: 100 * time.Millisecond,
```

Comment:
I think no need to wait longer, we had tested on csi snapshottor with same way.

Reply:
Then what's the point of waiting here? We can just exit here with `klog.Fatalf` if some CRDs are missed.

Reply:
we don't want to the container crash Immediately，if there is bad network,  so we retry some times.

Reply:
If there is a network issue, it will return an error in the check function above, and it will abort immediately, there is no chance to wait.

Reply:
If list the resource with errors that return (false, nil), is it ok?  retry sometimes.

Reply:
Is it needed to wait a moment when crds not deployed?

Reply:
I think it's OK not to wait here, WDYT? cc @liubog2008 @csuzhangxc

Reply:
wait or not wait are both OK for me.

Reply:
I think no need to crash immediately.

Reply:
I think "just let it crash" is more reasonable. @DanielZhangQD

Reply:
@zhucan Informer cache will wait after starting, no need to wait again.

Reply:
Use `ctx.Done` of timeout context to replace `wait.NeverStop` channel.

Reply:
@liubog2008 List the resources throught API server, not informer,  we retry it only wait crd deployed. In general, we prefer retrying with exponential backoff so that we don't potentially overload the API server with too many requests in a short time.

Reply:
however we'd like to have a faster startup time, if the container exits quickly and goes into CrashLoopBackOff it'll be restarted every 10s, 20s, 40s and so on, during the first run instead of failing quickly it might be possible that the CRDs are ready after a few seconds and in one of these 10 runs the wait condition might succeed @DanielZhangQD @liubog2008

Reply:
@zhucan I mean informer also will wait before CRD is deployed. It's no need to wait twice at begin.

Reply:
@liubog2008 yes, where is the better place to wait? do you have better suggestion?

Reply:
@zhucan See https://github.com/pingcap/tidb-operator/blob/master/cmd/controller-manager/main.go#L181
you can change `wait.NeverStop` to a context created by `context.WithTimeout()`

Reply:
I don't know whether it meets your requirement.